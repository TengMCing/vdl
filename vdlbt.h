//
// Created by Patrick Li on 5/10/22.
//

#ifndef VDL_VDLBT_H
#define VDL_VDLBT_H

#include "vdlutil.h"
#include <string.h>

/*-----------------------------------------------------------------------------
 |  Backtrace definition
 ----------------------------------------------------------------------------*/

#define VDL_STACK_FRAME_LIMIT 1000
#define VDL_FUNC_NAME_LIMIT 10000

/// @description Backtrace struct.
/// @param file_name (const char *) File name.
/// @param line_num (int) Line number.
typedef struct vdl_bt
{
    const char *const file_name;
    const int line_num;
} vdl_bt;

/// @description Global backtrace.
static struct
{
    int num_frame;
    int line_num[VDL_STACK_FRAME_LIMIT];
    const char *func_name[VDL_STACK_FRAME_LIMIT];
    const char *file_name[VDL_STACK_FRAME_LIMIT];
} VDL_GBT = {0};

/*-----------------------------------------------------------------------------
 |  Macro to work with backtrace
 ----------------------------------------------------------------------------*/

/// @description Push the function to the top of the backtrace stack.
/// @details This should be used at the beginning of every function that
/// needs to be traced.
#define vdl_PushBT(bt)                                         \
    do {                                                       \
        VDL_GBT.line_num[VDL_GBT.num_frame]  = (bt).line_num;  \
        VDL_GBT.func_name[VDL_GBT.num_frame] = __func__;       \
        VDL_GBT.file_name[VDL_GBT.num_frame] = (bt).file_name; \
        VDL_GBT.num_frame++;                                   \
    } while (0)

/// @description Remove the top frame from the backtrace stack and return the value.
/// @details The return value should not contain any further calls to functions.
#define vdl_ReturnConst(object) \
    do {                        \
        VDL_GBT.num_frame--;    \
        return object;          \
    } while (0)

/// @description Automatically push and pop the backtrace stack.
/// @details This macro should be used in every function needs to be traced.
// #define vdl_WithBT() for (int c = (VDL_GBT.line_num[VDL_GBT.num_frame]  = (bt).line_num,  \
//                                    VDL_GBT.func_name[VDL_GBT.num_frame] = __func__,       \
//                                    VDL_GBT.file_name[VDL_GBT.num_frame] = (bt).file_name, \
//                                    VDL_GBT.num_frame++,                                   \
//                                    0);                                                    \
//                           c < 1;                                                          \
//                           (VDL_GBT.num_frame--, c++))

#define vdl_bt_Call(func, ...) func((vdl_bt){.file_name = __FILE__, .line_num = __LINE__}, ##__VA_ARGS__)

/*-----------------------------------------------------------------------------
 |  Print backtrace
 ----------------------------------------------------------------------------*/

static inline const char *vdl_bt_Whitespaces(char *s, const int num)
{
    vdl_For_i(num) s[i] = ' ';
    s[num]              = '\0';
    return s;
}

static inline void vdl_bt_Print(void)
{
    printf("Backtrace - %d stack frames:\n", VDL_GBT.num_frame);
    char whitespaces[VDL_FUNC_NAME_LIMIT] = {'\0'};
    int func_length[VDL_GBT.num_frame];
    int max_func_length = -1;
    vdl_For_i(VDL_GBT.num_frame)
    {
        func_length[i] = (int) strlen(VDL_GBT.func_name[i]);
        if (func_length[i] > max_func_length)
            max_func_length = func_length[i];
    }

    // printf("  ║═[0] Starting at <main>\n");
    for (int i = VDL_GBT.num_frame - 1; i >= 0; i--)
    {
        if (i != 0)
            printf("  ║═");
        else
            printf("  ╚═");
        printf("[%d] Calling <%.*s>%s from %s:%d\n",
               VDL_GBT.num_frame - i - 1,
               (int) strlen(VDL_GBT.func_name[i]) - 3,
               VDL_GBT.func_name[i],
               vdl_bt_Whitespaces(whitespaces, max_func_length - (int) strlen(VDL_GBT.func_name[i])),
               VDL_GBT.file_name[i], VDL_GBT.line_num[i]);
    }
}

#endif//VDL_VDLBT_H
