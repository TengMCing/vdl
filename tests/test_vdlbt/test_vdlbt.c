//
// Created by Patrick Li on 23/10/22.
//

#include "../../vdlbt.h"
#include "../test.h"
#include <limits.h>

#define MAX_PATH_LENGTH 10000

static int foo_BT(vdl_bt bt);
#define foo() vdl_Caller(foo_BT, int)
static void foo_noreturn_BT(vdl_bt bt);
#define foo_noreturn() vdl_CallerNoReturn(foo_noreturn_BT)
static void foo_error_BT(vdl_bt bt);
#define foo_error() vdl_CallerNoReturn(foo_error_BT)

int main(void)
{
    // echo
    echo("Test vdl_MakeBT:");
    // expect(True)
    if (strncmp(__FILE__, vdl_MakeBT().file_name, MAX_PATH_LENGTH) == 0)
        test_printf("True");
    else
        test_printf("False");
    // expect(28)
    test_printf("%d", vdl_MakeBT().line_num);

    // echo
    echo("Test vdl_PushBT and vdl_PopBT:");
    foo();

    // defer expect
    test_printf("%d", VDL_GBT.num_frame);

    foo_error();

    return 0;
VDL_EXCEPTION:
    // exit(1)
    test_printf_wrapper(vdl_Abort());
}

static int foo_BT(vdl_bt bt)
{
    vdl_PushBT(bt);
    // expect(32)
    test_printf("%d", VDL_GBT.line_num[0]);
    // expect(1)
    test_printf("%d", VDL_GBT.num_frame);
    // expect(True)
    if (strncmp(__FILE__, VDL_GBT.file_name[0], MAX_PATH_LENGTH) == 0)
        test_printf("True");
    else
        test_printf("False");
    // expect(foo_BT)
    test_printf("%s", VDL_GBT.func_name[0]);
    foo_noreturn();
    return 0;
VDL_EXCEPTION:
    test_printf("foo_BT exception triggered!");
    return INT_MAX;
}

static void foo_noreturn_BT(vdl_bt bt)
{
    vdl_PushBT(bt);
    // expect(59)
    test_printf("%d", VDL_GBT.line_num[1]);
    // expect(2)
    test_printf("%d", VDL_GBT.num_frame);
    // expect(True)
    if (strncmp(__FILE__, VDL_GBT.file_name[1], MAX_PATH_LENGTH) == 0)
        test_printf("True");
    else
        test_printf("False");
    // expect(foo_noreturn_BT)
    test_printf("%s", VDL_GBT.func_name[1]);
    // expect start
    // Backtrace - 2 stack frames:
    //   ║═[0] Calling <foo_noreturn> from test_vdlbt/test_vdlbt.c:59
    //   ╚═[1] Calling <foo>          from test_vdlbt/test_vdlbt.c:32
    // expect end
    test_printf_wrapper(vdl_bt_Print());
    return;
VDL_EXCEPTION:
    test_printf("foo_noreturn_BT exception triggered!");
    return;
}

// Line 35
// expect(0)

static void foo_error_BT(vdl_bt bt)
{
    vdl_PushBT(bt);
    // expect(foo_error_BT before error!)
    test_printf("foo_error_BT before error!");
    if (1 != 0)
    {
        VDL_GERROR = 1;
        // expect start
        // Backtrace - 1 stack frames:
        //   ╚═[0] Calling <foo_error> from test_vdlbt/test_vdlbt.c:37
        // expect end
        test_printf_wrapper(vdl_bt_Print());
        goto VDL_EXCEPTION;
    }
    test_printf("foo_error_BT no error!");
    return;
VDL_EXCEPTION:
    // expect(foo_noreturn_BT exception triggered!)
    test_printf("foo_noreturn_BT exception triggered!");
    return;
}

// expect(Program aborted!)
