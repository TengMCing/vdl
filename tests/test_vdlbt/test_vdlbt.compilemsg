In file included from test_vdlbt/test_vdlbt.c:6:
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
In file included from test_vdlbt/test_vdlbt.c:5:
test_vdlbt/../../vdlbt.h:113:44: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        return_type _rv = func(vdl_MakeBT(), ##__VA_ARGS__); \
                                           ^
test_vdlbt/test_vdlbt.c:32:5: warning: use of GNU statement expression extension [-Wgnu-statement-expression]
    foo();
    ^
test_vdlbt/test_vdlbt.c:12:15: note: expanded from macro 'foo'
#define foo() vdl_Caller(foo_BT, int)
              ^
test_vdlbt/../../vdlbt.h:112:6: note: expanded from macro 'vdl_Caller'
    ({                                                       \
     ^
In file included from test_vdlbt/test_vdlbt.c:6:
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
In file included from test_vdlbt/test_vdlbt.c:5:
test_vdlbt/../../vdlbt.h:121:26: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        func(vdl_MakeBT(), ##__VA_ARGS__); \
                         ^
In file included from test_vdlbt/test_vdlbt.c:6:
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
In file included from test_vdlbt/test_vdlbt.c:5:
test_vdlbt/../../vdlbt.h:121:26: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        func(vdl_MakeBT(), ##__VA_ARGS__); \
                         ^
In file included from test_vdlbt/test_vdlbt.c:6:
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlbt/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
23 warnings generated.
