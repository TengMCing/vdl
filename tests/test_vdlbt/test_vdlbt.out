			Test vdl_MakeBT:
				Subtest 1: True
				Subtest 2: 28
			Test vdl_PushBT and vdl_PopBT:
				Subtest 1: 32
				Subtest 2: 1
				Subtest 3: True
				Subtest 4: foo_BT
				Subtest 5: 59
				Subtest 6: 2
				Subtest 7: True
				Subtest 8: foo_noreturn_BT
				Subtest 9: Backtrace - 2 stack frames:
  ║═[0] Calling <foo_noreturn> from test_vdlbt/test_vdlbt.c:59
  ╚═[1] Calling <foo>          from test_vdlbt/test_vdlbt.c:32
				Subtest 10: 0
				Subtest 11: foo_error_BT before error!
				Subtest 12: Backtrace - 1 stack frames:
  ╚═[0] Calling <foo_error> from test_vdlbt/test_vdlbt.c:37
				Subtest 13: foo_noreturn_BT exception triggered!
				Subtest 14: Program aborted!
